### PR TITLE
TUI Beta

### PR DESCRIPTION
**Pull request by apazzolini**
This PR implements a textual user interface for Doppler. It took heavy inspiration (both visually and in code) from [lazygit](https://github.com/jesseduffield/lazygit) and is launched via `doppler tui`.

It currently supports switching between projects and configs and allows you to filter for, view, edit, add, delete, and copy secrets. For now, the keybindings and colors are hardcoded, but we could add the ability for users to customize these in later versions. Pressing `?` will show a popup describing the available actions.

I've been using this for the last few months without issue, but this should be considered an **alpha** implementation until it's been used a bit more widely.

image
image

### CODE REVIEW

File: pkg/tui/common/logging.go
Comment (Piccirello) [inline pkg/tui/common/logging.go]: Remove this comment?

File: pkg/tui/gui/cmp_status.go
Comment (Piccirello) [inline pkg/tui/gui/cmp_status.go]: These new copyright notices should use 2023

File: pkg/cmd/tui.go
Comment (Piccirello) [inline pkg/cmd/tui.go]: Is there some way we can make it apparent that this is in alpha and this could disappear at any time?

**Review by Piccirello — DISMISSED** [review]
I have only reviewed the code outside of the `pkg/tui` directory. Very excited to get this out into the wild.

File: pkg/cmd/tui.go
Comment (apazzolini) [reply pkg/cmd/tui.go]: I changed the text to `Launch TUI (alpha - may change or disappear in the future)`

File: pkg/tui/gui/cmp_secret_view.go
Comment (nmanoogian) [inline pkg/tui/gui/cmp_secret_view.go]: Not sure where the bug is exactly but if you have an empty secret (no name or value), it panics when you try to save

**Review by nmanoogian — COMMENTED** [review]
Still reviewing but wanted to drop this comment

File: pkg/tui/gui/dispatch.go
Comment (nmanoogian) [inline pkg/tui/gui/dispatch.go:207]: When I save, it doesn't show the "Fetching..." + spinner like it does when I press enter on a project or config. Why is that?

File: pkg/cmd/tui.go
Comment (Piccirello) [reply pkg/cmd/tui.go]: Was thinking about this a bit more. Is there a phase in which we'll transition this from alpha to beta before GA? If not, I think we can label this command `Launch TUI (BETA)`.

File: pkg/tui/gui/cmp_secret_view.go
Comment (apazzolini) [reply pkg/tui/gui/cmp_secret_view.go]: Good find. Fixed.

File: pkg/tui/gui/gui.go
Comment (nmanoogian) [inline pkg/tui/gui/gui.go:25]: curious: Why are we using this mutex instead of a standard RWMutex?

File: pkg/tui/gui/gui.go
Comment (nmanoogian) [inline pkg/tui/gui/gui.go]: recommendation: What do you think about moving this print into the default branch so it doesn't print "quit" on normal exits?

```diff
diff --git a/pkg/tui/gui/gui.go b/pkg/tui/gui/gui.go
index 653e50e..0c58387 100644
--- a/pkg/tui/gui/gui.go
+++ b/pkg/tui/gui/gui.go
@@ -159,11 +159,11 @@ func recoverScreenOnCrash() {
 func (gui *Gui) RunAndHandleError() error {
 	return SafeWithError(func() error {
 		if err := gui.run(); err != nil {
-			fmt.Println(err.Error())
 			switch err {
 			case gocui.ErrQuit:
 				return nil
 			default:
+				fmt.Println(err.Error())
 				return err
 			}
 		}


```

**Review by nmanoogian — DISMISSED** [review]
So excited to see this go out!

File: pkg/tui/gui/dispatch.go
Comment (apazzolini) [reply pkg/tui/gui/dispatch.go:207]: Good catch. The save function wasn't invoked in a goroutine.

File: pkg/tui/gui/gui.go
Comment (apazzolini) [reply pkg/tui/gui/gui.go:25]: I believe it's because this one does some amount of deadlock detection. Lazygit uses this one, so I followed suit.

**Comment by apazzolini** [comment]
All feedback is addressed, re-requesting reviews for final approval. @Piccirello will defer to you for timing re:merging.

**Review by Piccirello — DISMISSED** [review]
Once approved by Nic, merge at will

**Review by nmanoogian — DISMISSED** [review]

**Comment by apazzolini** [comment]
@Piccirello re-requesting one more time; had to rebase and resolve some Semgrep conflicts. I also went ahead and added arrow key bindings for navigating around the secret view.

**Review by Piccirello — DISMISSED** [review]

File: pkg/tui/gui/cmp_configs.go
Comment (nmanoogian) [inline pkg/tui/gui/cmp_configs.go:65]: Is this logic correct? If `err` is non-nil, it doesn't look like we ever return it

File: pkg/tui/gui/cmp_projects.go
Comment (nmanoogian) [inline pkg/tui/gui/cmp_projects.go:65]: Similar logic question

**Review by nmanoogian — DISMISSED** [review]
Updates look good! Can you rename all commits with `chore:` that you don't want to appear in the changelog?

File: pkg/tui/gui/cmp_configs.go
Comment (apazzolini) [reply pkg/tui/gui/cmp_configs.go:65]: Good catch. Fixed.

File: pkg/tui/gui/cmp_projects.go
Comment (apazzolini) [reply pkg/tui/gui/cmp_projects.go:65]: Good catch. Fixed.

**Comment by apazzolini** [comment]
> Updates look good! Can you rename all commits with `chore:` that you don't want to appear in the changelog?

Done

**Comment by Piccirello** [comment]
You'll probably want to prepend `chore:` to every commit other than "Add Doppler TUI" since they'll get added to the changelog. And maybe reword that to "Add tui command for editing secrets" or something like that.

**Comment by apazzolini** [comment]
> You'll probably want to prepend `chore:` to every commit other than "Add Doppler TUI" since they'll get added to the changelog. And maybe reword that to "Add tui command for editing secrets" or something like that.

Done

**Review by Piccirello — APPROVED** [review]

**Review by nmanoogian — APPROVED** [review]

**Pull request merged**
